### PR TITLE
Fix console for ie8

### PR DESCRIPTION
**Pull request by unional**
IE8 typeof returns "object" instead of "function", so checking for "function" doesn't work.

### CODE REVIEW

**Comment by scothis** [comment]
this is one of my personal favorite WTFIEs

**Comment by briancavalier** [comment]
@unional Thanks!  Is it safe to use `typeof localConsole.log !== 'undefined'` ?

**Comment by sompylasar** [comment]
There are more issues that may arise in IE8. Just accessing a property on the `console` object may throw an exception (e.g. `typeof console.log` may throw). Everything around the console object, including property access and method calls, should be wrapped in `try..catch` blocks.

**Comment by unional** [comment]
@briancavalier I think it is ok to do `typeof localConsole.log !== 'undefined'`, since this really only affect ie <=8 and quirks mode. I will update the code accordingly.
Also, I'll update the code to better support scenario without JSON. Currently, it will throw error in `ConsoleReporter.log`.

**Comment by briancavalier** [comment]
@unional [when.js 3.x requires an ES5 environment](https://github.com/cujojs/when/blob/master/docs/installation.md#getting-started), and since [JSON is part of the ES5 spec](https://es5.github.io/x15.12.html), we can assume it's there and works.  Folks needing to support older browsers can polyfill to ES5.

In contrast, `console` isn't part of EcmaScript.  So, I'm cool with jumping through a few more hoops to make sure we're not breaking envs that have console quirks, like IE.

**Comment by briancavalier** [comment]
@unional looking good.  If you squash the commits, I'll get this merged asap!

File: monitor/ConsoleReporter.js [hunk 1]
```diff
@@ -72,16 +72,20 @@ define(function(require) {
 				// IE8 has console.log and JSON, so we can make a
 				// reasonably useful warn() from those.
 				// Credit to webpro (https://github.com/webpro) for this idea
-				if (typeof localConsole.log ==='function'
-					&& typeof JSON !== 'undefined') {
-					log = warn = function (x) {
-						if(typeof x !== 'string') {
+				// typeof localConsole.log will return 'object' in IE8, so can't test it with === 'function'
+				// Since this is more of a corner case for IE8, I'm ok to check it with !== 'undefined' to reduce complexity
+				if (typeof localConsole.log !== 'undefined' && typeof JSON !== 'undefined') {
```
Comment (sompylasar) [inline monitor/ConsoleReporter.js]: Just **accessing a property** on the `console` object (`localConsole.log` here and `localConsole.error`, `localConsole.dir` above) may **throw an exception** in IE8, and the whole library will terminate its execution.

Comment (unional) [inline monitor/ConsoleReporter.js]: Does it fail because console is not available in IE8? The code already checked for that so if that is the cause, the code is fine. 

Homa
Increase the Net Worth of your time

Comment (sompylasar) [inline monitor/ConsoleReporter.js]: The code that uses only `typeof` would still fail because the execution internally goes in the following steps:
1. Access `localConsole.error`, [the line where it happens is in the collapsed code above, not visible in this PR by default](https://github.com/cujojs/when/pull/436/files#diff-2eb9f8e6caaab86409cabd1f4c79cac7R53). **While performing this operation IE8 can throw** sort of `Access denied.` exception AFAIR.
2. Apply the `typeof` operator to the result of the expression evaluated above. **This second step won't even get executed** if the previous one throws an exception.

Comment (briancavalier) [inline monitor/ConsoleReporter.js]: @sompylasar Do you have a link to some info on that?  I'm not aware of a case where IE would throw on `console` property access (unless of course, `console` is undefined, which is already covered).  I certainly wouldn't put it past IE doing something crazy :D but I'd like to verify.  Thanks!

Comment (sompylasar) [inline monitor/ConsoleReporter.js]: @briancavalier I'm sorry this is my own experience, I don't have a link and I have found no traces about this issue on the Internet. I recall that this happened in certain debugging conditions: I had IE11 in IE8 mode with the console tab not opened initially, and there was some navigation that likely broke the connection between the webpage and the console object (like the closured console instance got destroyed internally and was not substituted by the new one). Maybe this is an issue in the Developer Tools because in IE the `console` object is provided by them. But I saw similar `Access denied` (COM-related) issues for `window` and `location` objects in production; they popped out if references to `window` and `location` were stored in a closure and later accessed after the `document.domain` changed. So I cannot guarantee this won't happen with the `console` object because it's COM-based, too, thus fragile. It is important to make these helper objects cope with any error happening inside of them not to break the surrounding code.

Comment (briancavalier) [inline monitor/ConsoleReporter.js]: @sompylasar Thanks for the explanation.  It's def a concerning behavior, but I also haven't found any evidence of it after quite a few google searches.

I think the best thing to do is to proceed with this PR as is, and if/when we find evidence that need to be more paranoid and add `try/catch`, we can add it at that point.

**Comment by briancavalier** [comment]
@unional Ping.  I'm happy to merge this if you'll squash the commits.  Thanks!

**Comment by unional** [comment]
@brian, sorry for the late reply. Have been busy lately. I have attempted
to squashed the commits. I don't know if I'm doing it correctly. If I
wasn't, please let me know and I will redo it.

Thanks,

**Comment by briancavalier** [comment]
@unional Thanks.  There's a merge commit in the mix now, unfortunately.  No worries, though, I'll merge your change in 168a26d manually.

**Comment by briancavalier** [comment]
@unional Sorry for letting this sit for a while!  I just merged your change into master in bfee33d.  I'll pull together a new release later this week.

Thanks again!
